Test failures for sparse sum tests on 32-bit Python

Getting a large amount of test failures like the one below, I assume that this is due to gh-5907:

    ======================================================================
    FAIL: test_base.TestCSC.test_sum(<type 'numpy.uint64'>, 1)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
        self.test(*self.arg)
      File "/home/rgommers/Code/scipy/scipy/sparse/tests/test_base.py", line 767, in check
        assert_equal(dat.sum(axis=0).dtype, datsp.sum(axis=0).dtype)
      File "/home/rgommers/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 375, in assert_equal
        raise AssertionError(msg)
    AssertionError: 
    Items are not equal:
     ACTUAL: dtype('uint64')
     DESIRED: dtype('uint32')

And also a large set of overflow warnings, all on the same line:

    /home/rgommers/Code/scipy/scipy/sparse/dok.py:370: RuntimeWarning: overflow encountered in ulonglong_scalars
      result[i] += v * other[j]

--- a/scipy/sparse/base.py
+++ b/scipy/sparse/base.py
@@ -757,10 +757,11 @@ class spmatrix(object):
         # Mimic numpy's casting.
         if np.issubdtype(self.dtype, np.float_):
             res_dtype = np.float_
-        elif self.dtype.kind in 'ib':
-            res_dtype = np.int_
-        elif self.dtype.kind == 'u':
+        elif (self.dtype.kind == 'u' and
+              np.can_cast(self.dtype, np.uint)):
             res_dtype = np.uint
+        elif np.can_cast(self.dtype, np.int_):
+            res_dtype = np.int_
         else:
             res_dtype = self.dtype
 

--- a/scipy/sparse/compressed.py
+++ b/scipy/sparse/compressed.py
@@ -567,10 +567,11 @@ class _cs_matrix(_data_matrix, _minmax_mixin, IndexMixin):
             # Mimic numpy's casting.
             if np.issubdtype(self.dtype, np.float_):
                 res_dtype = np.float_
-            elif self.dtype.kind in 'ib':
-                res_dtype = np.int_
-            elif self.dtype.kind == 'u':
+            elif (self.dtype.kind == 'u' and
+                  np.can_cast(self.dtype, np.uint)):
                 res_dtype = np.uint
+            elif np.can_cast(self.dtype, np.int_):
+                res_dtype = np.int_
             else:
                 res_dtype = self.dtype
             ret = np.zeros(len(self.indptr) - 1, dtype=res_dtype)

--- a/scipy/sparse/tests/test_base.py
+++ b/scipy/sparse/tests/test_base.py
@@ -758,19 +758,22 @@ class _TestCommon:
         def check(dtype, j):
             dat = np.matrix(matrices[j], dtype=dtype)
             datsp = self.spmatrix(dat, dtype=dtype)
-            assert_array_almost_equal(dat.sum(), datsp.sum())
-            assert_equal(dat.sum().dtype, datsp.sum().dtype)
-            assert_(np.isscalar(datsp.sum(axis=None)))
-            assert_array_almost_equal(dat.sum(axis=None), datsp.sum(axis=None))
-            assert_equal(dat.sum(axis=None).dtype, datsp.sum(axis=None).dtype)
-            assert_array_almost_equal(dat.sum(axis=0), datsp.sum(axis=0))
-            assert_equal(dat.sum(axis=0).dtype, datsp.sum(axis=0).dtype)
-            assert_array_almost_equal(dat.sum(axis=1), datsp.sum(axis=1))
-            assert_equal(dat.sum(axis=1).dtype, datsp.sum(axis=1).dtype)
-            assert_array_almost_equal(dat.sum(axis=-2), datsp.sum(axis=-2))
-            assert_equal(dat.sum(axis=-2).dtype, datsp.sum(axis=-2).dtype)
-            assert_array_almost_equal(dat.sum(axis=-1), datsp.sum(axis=-1))
-            assert_equal(dat.sum(axis=-1).dtype, datsp.sum(axis=-1).dtype)
+            with np.errstate(over='ignore'):
+                assert_array_almost_equal(dat.sum(), datsp.sum())
+                assert_equal(dat.sum().dtype, datsp.sum().dtype)
+                assert_(np.isscalar(datsp.sum(axis=None)))
+                assert_array_almost_equal(dat.sum(axis=None),
+                                          datsp.sum(axis=None))
+                assert_equal(dat.sum(axis=None).dtype,
+                             datsp.sum(axis=None).dtype)
+                assert_array_almost_equal(dat.sum(axis=0), datsp.sum(axis=0))
+                assert_equal(dat.sum(axis=0).dtype, datsp.sum(axis=0).dtype)
+                assert_array_almost_equal(dat.sum(axis=1), datsp.sum(axis=1))
+                assert_equal(dat.sum(axis=1).dtype, datsp.sum(axis=1).dtype)
+                assert_array_almost_equal(dat.sum(axis=-2), datsp.sum(axis=-2))
+                assert_equal(dat.sum(axis=-2).dtype, datsp.sum(axis=-2).dtype)
+                assert_array_almost_equal(dat.sum(axis=-1), datsp.sum(axis=-1))
+                assert_equal(dat.sum(axis=-1).dtype, datsp.sum(axis=-1).dtype)
 
         for dtype in self.checked_dtypes:
             for j in range(len(matrices)):

